libs/ape-core/ape/core/v2/trainer/base.py

```diff
@@ -0,0 +1,205 @@
+from abc import ABC, abstractmethod
+import asyncio
+import inspect
+import json
+import random
+from typing import Dict, Any, List, Optional, Tuple
+
+from ape.common.prompt.prompt_base import Prompt
+from ape.common.types import GlobalMetricResult, MetricResult, DatasetItem
+from ape.common.generate import BaseGenerate
+from ape.common.global_metric import BaseGlobalMetric
+from ape.common.metric import BaseMetric
+from ape.core.core_prompts import ApeCorePrompts
+from ape.core.v2.types.report import BaseReport
+
+class BaseTrainer(ABC):
+    def __init__(
+        self,
+        generator: BaseGenerate,
+        metric: BaseMetric,
+        global_metric: Optional[BaseGlobalMetric] = None,
+        task_description: Optional[str] = None,
+        metric_description: Optional[str] = None,
+        **kwargs,
+    ):
+        self.generator = generator
+        self.metric = metric
+        self.global_metric = global_metric
+        self.task_description = task_description
+        self.metric_description = metric_description
+        self.dataset_summary = None
+        
+    @abstractmethod
+    async def fit(
+        self,
+        prompt: Prompt,
+        trainset: List[DatasetItem],
+        valset: List[DatasetItem],
+    ) -> Tuple[Prompt, BaseReport]:
+        """
+        Train model
+
+        Args:
+            prompt (Prompt): Prompt
+            trainset (List[DatasetItem]): Training dataset
+            valset (List[DatasetItem]): Validation dataset
+
+        Returns:
+            Tuple[Prompt, BaseReport]: Trained prompt and report
+        """
+        pass
+
+    async def __call__(self, prompt: Prompt, trainset: List[DatasetItem], valset: List[DatasetItem]) -> Tuple[Prompt, BaseReport]:
+        return await self.train(prompt, trainset, valset)
+    
+    async def _evaluate_dataset(
```

Comment:
1) To me the name of this method is a bit tricky - it's not evaluating a dataset, it's using a dataset for evaluation
2) Maybe move this to a util module and use it from `Evaluate` as well

Reply:
move it to util mode and use it from `Evaluate` class.
Recommended purpose of `Evaluate` class is saving `(dataset, generate, metric, global_metric)` tuple. 
So when initiating, `dataset: List[DatasetItem]` must be passed. 
However, to allow flexible use, it is OK to allow users to pass new dataset when call `compute`.

Reply:
`_evaluate_dataset` -> `_evaluate`

Reply:
not resolved yet